<a href="https://colab.research.google.com/github/2303A52060/sml_assignments/blob/main/SML_A_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Assignment-2:

Given the following dataset: X={1.3,1.5,2.6,2.6,3.2,3.9,4.2,3.7,3.10,3.0,11,6.7,1.9}. Compute the density of the data using a histogram with 5 bins.

Question 2:
dependent variable:
12,13,56,23,41,28,51
independent variable:
51,63,37,88,33,72,81

Find Mean, Median, Mode , Variance , Standard deviation , Range , Interquartile Range (IQR),  Skewness ,Kurtosis of x and y .

Question 3:

From the above data Find Mean, Median, Mode, Variance, Standard deviation, Range , Interquartile Range (IQR),  Skewness ,Kurtosis of each feature.


In [ ]:
import numpy as np
import pandas as pd
from scipy import stats

# Dataset for Question 1
X = np.array([1.3, 1.5, 2.6, 2.6, 3.2, 3.9, 4.2, 3.7, 3.10, 3.0, 11, 6.7, 1.9])

# Define bins
num_bins = 5
bin_edges = np.linspace(X.min(), X.max(), num_bins + 1)

# Compute histogram
hist, bin_edges = np.histogram(X, bins=bin_edges)
bin_width = bin_edges[1] - bin_edges[0]
density = hist / (bin_width * len(X))

# Print density results
print("Question 1: Histogram Density")
for i in range(num_bins):
    print(f"Bin {i + 1}: {bin_edges[i]:.2f} to {bin_edges[i + 1]:.2f}, Density: {density[i]:.4f}")

# Data for Question 2
dependent = np.array([12, 13, 56, 23, 41, 28, 51])
independent = np.array([51, 63, 37, 88, 33, 72, 81])

# Function to compute statistics
def compute_statistics(data):
    mean = np.mean(data)
    median = np.median(data)
    mode_result = stats.mode(data)
    mode = mode_result.mode  # Extract the mode value
    variance = np.var(data, ddof=0)
    std_dev = np.std(data, ddof=0)
    data_range = np.ptp(data)
    iqr = stats.iqr(data)
    skewness = stats.skew(data)
    kurtosis = stats.kurtosis(data)

    return {
        'Mean': mean,
        'Median': median,
        'Mode': mode,
        'Variance': variance,
        'Standard Deviation': std_dev,
        'Range': data_range,
        'IQR': iqr,
        'Skewness': skewness,
        'Kurtosis': kurtosis
    }

# Compute statistics for dependent and independent variables
stats_dependent = compute_statistics(dependent)
stats_independent = compute_statistics(independent)

# Print results for Question 2
print("\nQuestion 2: Statistics for Dependent Variable")
for stat_name, value in stats_dependent.items():
    print(f"{stat_name}: {value:.4f}")

print("\nQuestion 2: Statistics for Independent Variable")
for stat_name, value in stats_independent.items():
    print(f"{stat_name}: {value:.4f}")

# Data for Question 3 (same as above data)
# For each feature in the dataset, the statistics computed are from Question 2
# We already have these statistics computed.

print("\nQuestion 3: Statistics for Each Feature (same as above data)")
print("Dependent Variable Statistics:")
for stat_name, value in stats_dependent.items():
    print(f"{stat_name}: {value:.4f}")

print("\nIndependent Variable Statistics:")
for stat_name, value in stats_independent.items():
    print(f"{stat_name}: {value:.4f}")



Question 1: Histogram Density
Bin 1: 1.30 to 3.24, Density: 0.3172
Bin 2: 3.24 to 5.18, Density: 0.1190
Bin 3: 5.18 to 7.12, Density: 0.0397
Bin 4: 7.12 to 9.06, Density: 0.0000
Bin 5: 9.06 to 11.00, Density: 0.0397

Question 2: Statistics for Dependent Variable
Mean: 32.0000
Median: 28.0000
Mode: 12.0000
Variance: 268.0000
Standard Deviation: 16.3707
Range: 44.0000
IQR: 28.0000
Skewness: 0.1876
Kurtosis: -1.4768

Question 2: Statistics for Independent Variable
Mean: 60.7143
Median: 63.0000
Mode: 33.0000
Variance: 387.6327
Standard Deviation: 19.6884
Range: 55.0000
IQR: 32.5000
Skewness: -0.1016
Kurtosis: -1.4265

Question 3: Statistics for Each Feature (same as above data)
Dependent Variable Statistics:
Mean: 32.0000
Median: 28.0000
Mode: 12.0000
Variance: 268.0000
Standard Deviation: 16.3707
Range: 44.0000
IQR: 28.0000
Skewness: 0.1876
Kurtosis: -1.4768

Independent Variable Statistics:
Mean: 60.7143
Median: 63.0000
Mode: 33.0000
Variance: 387.6327
Standard Deviation: 19.6884
Range: 

solution:

Question 1: Compute the density of the data using a histogram with 5 bins
Given dataset:

𝑋
=
{
1.3
,
1.5
,
2.6
,
2.6
,
3.2
,
3.9
,
4.2
,
3.7
,
3.10
,
3.0
,
11
,
6.7
,
1.9
}
X={1.3,1.5,2.6,2.6,3.2,3.9,4.2,3.7,3.10,3.0,11,6.7,1.9}.

Determine the Range:

Minimum value:
1.3
1.3

Maximum value:
11
11

Range:
11
−
1.3
=
9.7
11−1.3=9.7

Determine Bin Width:

Number of bins: 5
Bin width:
9.7
5
=
1.94
≈
1.94
5
9.7
​
 =1.94≈1.94

Create Bins:

Bin 1:
[
1.3
,
3.24
)
[1.3,3.24)
Bin 2:
[
3.24
,
5.18
)
[3.24,5.18)
Bin 3:
[
5.18
,
7.12
)
[5.18,7.12)
Bin 4:
[
7.12
,
9.06
)
[7.12,9.06)
Bin 5:
[
9.06
,
11
)
[9.06,11)

Count the Data Points in Each Bin:

Bin 1:
{
1.3
,
1.5
,
2.6
,
2.6
,
1.9
,
3.0
}
{1.3,1.5,2.6,2.6,1.9,3.0} → 6 data points

Bin 2:
{
3.2
,
3.9
,
3.7
}
{3.2,3.9,3.7} → 3 data points

Bin 3:
{
4.2
}
{4.2} → 1 data point
Bin 4:

{
6.7
}
{6.7} → 1 data point

Bin 5:
{
11
}
{11} → 1 data point
Compute Density:

Total number of data points: 13

Bin densities (assuming equal bin width):

Density for Bin 1
=
6
13
×
1.94
≈
0.243
Density for Bin 1=
13×1.94
6
​
 ≈0.243

Density for Bin 2
=
3
13
×
1.94
≈
0.118
Density for Bin 2=
13×1.94
3
​
 ≈0.118

Density for Bin 3
=
1
13
×
1.94
≈
0.031
Density for Bin 3=
13×1.94
1
​
 ≈0.031

Density for Bin 4
=
1
13
×
1.94
≈
0.031
Density for Bin 4=
13×1.94
1
​
 ≈0.031

Density for Bin 5
=
1
13
×
1.94
≈
0.031
Density for Bin 5=
13×1.94
1
​
 ≈0.031
Question 2: Compute Statistical Measures for Dependent and Independent Variables

Dependent Variable:
𝑌
=
{
12
,
13
,
56
,
23
,
41
,
28
,
51
}
Y={12,13,56,23,41,28,51}

Independent Variable:
𝑋
=
{
51
,
63
,
37
,
88
,
33
,
72
,
81
}
X={51,63,37,88,33,72,81}

1. Dependent Variable Statistics:

Mean:
12
+
13
+
56
+
23
+
41
+
28
+
51
7
=
224
7
=
32
7
12+13+56+23+41+28+51
​
 =
7
224
​
 =32

Median: Sort data:
{
12
,
13
,
23
,
28
,
41
,
51
,
56
}
{12,13,23,28,41,51,56} → Median = 28

Mode: No repeating values, so no mode.

Variance:

Variance
=
∑
(
𝑥
𝑖
−

mean
)
2
𝑁
=
(
12
−
32
)
2
+
(
13
−
32
)
2
+
…
+
(
51
−
32
)
2
7

Variance=
N
∑(x
i
​
 −mean)
2

​
 =
7
(12−32)
2
 +(13−32)
2
 +…+(51−32)
2

​


=
400
+
361
+
576
+
16
+
81
+
169
+
361
7
=
1964
7
≈
280.57
=
7
400+361+576+16+81+169+361
​
 =
7
1964
​
 ≈280.57

Standard Deviation:
280.57
≈
16.75
280.57
​
 ≈16.75

Range:
56
−
12
=
44
56−12=44

Interquartile Range (IQR):
Q1 = 23
Q3 = 51
IQR = Q3 - Q1 = 51 - 23 = 28

Skewness: 0.1876
Kurtosis: -1.4768
2. Independent Variable Statistics:

Mean:
51
+
63
+
37
+
88
+
33
+
72
+
81
7
=
445
7
=
63.57
7
51+63+37+88+33+72+81
​
 =
7
445
​
 =63.57

Median: Sort data:
{
33
,
37
,
51
,
63
,
72
,
81
,
88
}
{33,37,51,63,72,81,88} → Median = 63

Mode: No repeating values, so no mode.

Variance:
Variance
=
∑
(
𝑥
𝑖
−

mean
)
2
𝑁
=
(
51
−
63.57
)
2
+
(
63
−
63.57
)
2
+
…
+
(
88
−
63.57
)
2
7

Variance=
N
∑(x
i
​
 −mean)
2

​
 =
7
(51−63.57)
2
 +(63−63.57)
2
 +…+(88−63.57)
2

​


=
161.47
+
0.32
+
160.61
+
603.61
+
927.37
+
129.87
+
613.67
7
≈
398.60
=
7
161.47+0.32+160.61+603.61+927.37+129.87+613.67
​
 ≈398.60

Standard Deviation:
398.60
≈
19.97
398.60
​
 ≈19.97

Range:
88
−
33
=
55
88−33=

Interquartile Range (IQR):
Q1 = 51
Q3 = 81
IQR = Q3 - Q1 = 81 - 51 = 30

Skewness: -0.1016
Kurtosis: -1.4265

Question 3: Compute the same statistics for each feature
Since the dataset provided in Question 2 already provides features (dependent and independent variables),

Question 3:

From the above data Find Mean, Median, Mode, Variance, Standard deviation, Range , Interquartile Range (IQR),  Skewness ,Kurtosis of each feature.

solution:

In 2nd question code is same as question no:3,
already done in 2nd solution.